### ISSUE TITLE
Url parameter access on formRequest does not work

### ISSUE THREAD

**Issue by oskobri** [knuckleswtf/scribe#594]
Hello,

In a FormRequest, access directly to url parameter does not work (null). Here is a code example:

```
<?php

namespace App\Http\Requests;

use Illuminate\Foundation\Http\FormRequest;
use Illuminate\Support\Facades\Route;

class UserRequest extends FormRequest
{

    public function rules(): array
    {
        $this->user; // null

        $this->route('user'); // null

        Route::current()->parameters()['user']; // not null

    }
}
```

Only the third one works but i prefer to avoid this.

I ran `php artisan scribe:generate`

**Comment by Dieter91** [comment]
Ran into the same problem here. Third option does indeed work.

**Comment by MissaelAnda** [comment]
Same here but not even the third option works.
Im using ver. 4.12.0

**Comment by shalvah** [comment]
Thanks for all the reports. Help fixing this would also be appreciated.

**Comment by AunCly** [comment]
Hi, 

Same problem here i have a basic formRequest with the following rule : 
```php
Rule::unique('users', 'login')->ignore($this->user->login, 'login')
```

And scribe throw exception cause $this->user is null

I have dived into your package and started to write a solution. My solution work but only if i specify an @urlParam

For example : route PUT /api/users/{user} and controller : 
```php
/**
* @urlParam user int required L'identifiant de l'utilisateur. Exemple: 300
*/
public function update(UserRequest $request, UserModel $user){ // ... }
```

And here the following update that i have wrote into scribe 
Knuckles\Scribe\Extracting\Strategies\GetFromFormRequestBase.php

```php
    public function __invoke(ExtractedEndpointData $endpointData, array $routeRules = []): ?array
    {
        return $this->getParametersFromFormRequest($endpointData, $endpointData->method, $endpointData->route);
    }

    public function getParametersFromFormRequest(ExtractedEndpointData $endpointData, ReflectionFunctionAbstract $method, Route $route): array
    {
        if (!$formRequestReflectionClass = $this->getFormRequestReflectionClass($method)) {
            return [];
        }
        
        if (!$this->isFormRequestMeantForThisStrategy($formRequestReflectionClass)) {
            return [];
        }
        
        $className = $formRequestReflectionClass->getName();
        
        /*
         * Get the params name/example value 
         */
        $params = [];
        foreach($endpointData->urlParameters as $param) {
            $params[$param->name] = $param->example;
        }
    
        /**
         * instanciate a new request with the route parameters
         */
        $request = \Illuminate\Http\Request::create($route->uri(), $route->methods()[0], $params);
        
        if (Globals::$__instantiateFormRequestUsing) {
            $formRequest = call_user_func_array(Globals::$__instantiateFormRequestUsing, [$className, $route, $method]);
        } else {
            /**
             * instanciate a new form request
             */
            $formRequest = \Illuminate\Foundation\Http\FormRequest::createFrom($request, new $className);
        }
    
        /**
         * bind the route parameters to the form request
         */
        $route->bind($formRequest);
    
        /**
         * set the route parameters to the route
         */
        foreach($endpointData->urlParameters as $param) {
            $route->setParameter($param->name, $param->example);
        }
    
        /**
         * convert params to model instances
         */
        app('router')->substituteBindings($route);
        app('router')->substituteImplicitBindings($route);
        
        $formRequest->server->set('REQUEST_METHOD', $route->methods()[0]);
        
        $parametersFromFormRequest = $this->getParametersFromValidationRules(
            $this->getRouteValidationRules($formRequest),
            $this->getCustomParameterData($formRequest)
        );
        
        return $this->normaliseArrayAndObjectParameters($parametersFromFormRequest);
    }
```

But i have found 2 problems and thats why i don't PR your my solution.
First :  $endpointData->urlParameters have 2 keys, the one added by Strategies\UrlParameters\GetFromLaravelAPI::class and the second my @urlParams added by Strategies\UrlParameters\GetFromUrlParamTag::class, that don't cause problem but it's little strange that my @urlParams doesn't override the first one ?

Second : in the case i don't specify the @urlParams i have an not found exception throw by model binding cause GetFromLaravelAPI create a tag user_id and substitution doesn't work.

Hope it help. :)

**Comment by shalvah** [comment]
I've not looked at this in detail yet, but if you're unhappy with the results of GetFromLaravelAPI and want to adjust it, you can either:
1. Remove the strategy from your config to disable it entirely, OR
2. Use the `normalizeEndpointUsing()` hook to customise or disable URL normalization (or converting `/{user}` to `/{user_id}`. There are examples in the docs - https://scribe.knuckles.wtf/laravel/hooks#normalizeendpointurlusing

**Comment by AunCly** [comment]
I have already think to that, in fact i have no problem with this strategies, i just wanted to warn you about my fix problems, maybe i wasn't clear enough

**Comment by MissaelAnda** [comment]
Perhaps we can move `normalizeEndpointUsing()` to after extracting all data, and it would be nice if `ExtractedEndpointData` marked the `urlParams` example to Eloquent model binding so when extracting the body params it could try to do the models stratigies `databaseFirst`, `factoryCreate` or `factoryMake`.

**Comment by mortenscheel** [comment]
I had a similar issue, which I was able to solve by using [instantiateFormRequestUsing](https://scribe.knuckles.wtf/laravel/hooks#instantiateformrequestusing).

Here's a simple example where the rules depend on a route param:
```php
class SomeFormRequest extends FormRequest
{
    public Customer $customer;

    protected function prepareForValidation()
    {
        $this->customer = $this->route('customer');
    }

    public function rules()
    {
        return [
            'user_id' => ['required', Rule::exists('users', 'id')->where('customer_id', $this->customer->id)],
        ];
    }
}
```
And in AppServiceProvider's boot method:
```php
if (class_exists(Scribe::class)) {
    Scribe::instantiateFormRequestUsing(function (string $formRequestClassName) {
        if ($formRequestClassName === SomeSpecificFormRequest::class) {
            $formRequest = new $formRequestClassName();
            $formRequest->customer = new Customer();
        }

        return $formRequest;
    });
}
```
Note that the Scribe documentation's example uses `app()->makeWith()` to instantiate the formrequest, which doesn't work in my experience. When a FormRequest is resolved this way, it automatically triggers the `ValidatesWhenResolved` trait, meaning that a ValidationException is thrown if the rules don't pass (which they probably won't since there are no request params). But you _can_ instantiate the FormRequest using `new`, and then add "mock" dependencies, before passing it back to Scribe.

So basically this workaround works by having the dependencies defined as class attributes.

**Comment by gabrielrbarbosa** [comment]
@mortenscheel Thank you!! Using your solution solved [#https://github.com/knuckleswtf/scribe/issues/526](https://github.com/knuckleswtf/scribe/issues/526)

**Comment by patrickomeara** [comment]
I've just hit this issue after installing Scribe for the first time. I will look into a fix for this.

**Comment by patrickomeara** [comment]
I believe that in order to fix this issue, the path, annotated, attribute and method arguments should be merged and iterated over. If the types extend Model then instantiateExampleModel should be used to set the parameter on the route after it is bound using the request in `getParametersFromFormRequest`.

I'm still getting to know the codebase, for anyone experienced with the codebase does this sound like a suitable solution? 

From what I can see the models are only instantiated for responses, would instantiating them for typed method arguments be a possibility?

**Comment by shalvah** [comment]
Not gonna lie, I'm thoroughly confused here...is the problem model binding or something else?

**Comment by patrickomeara** [comment]
Correct, model binding isn't working in `rules()` 

I've added a test [here](https://github.com/patrickomeara/scribe/commit/b8148c4f617711a417db5fda2a083b2165d12e34)

When in the `rules()` method. `$this->route('post')`, `$this->post` and `Route::current()->parameters()['post']` should all return a post model. The first two return null, the latter returns a string.

**Comment by shalvah** [comment]
Ah, okay. Tbh, you have a better chance of fixing this than I do at the moment. 😅 Don't use Laravel much these days, and not been a fan of model binding tbh, especially within validation (and I'm also on vacation). PRs are welcome.

**Comment by igorsgm** [comment]
I'm also with the same issue. Any luck here?

**Comment by patrickomeara** [comment]
Hey @igorsgm I haven't looked at this for a while, but it was a sticking point for us. 

Have you tried the solution @mortenscheel mentioned?

**Comment by igorsgm** [comment]
Yes, I tried, but the model object is not bound in the request object. 

So I managed to achieve what I was looking for by extending GetFromLaravelAPI urlParameter strategy, because this strategy already infers the eloquent model parameter:

```php
use Knuckles\Camel\Extraction\ExtractedEndpointData;
use Knuckles\Scribe\Extracting\Shared\UrlParamsNormalizer;
use Knuckles\Scribe\Extracting\Strategies\UrlParameters\GetFromLaravelAPI;
use Throwable;

class GetFromLaravelAPIExtended extends GetFromLaravelAPI
{
    public function inferBetterTypesAndExamplesForEloquentUrlParameters(array $parameters, ExtractedEndpointData $endpointData): array
    {
        //We'll gather Eloquent model instances that can be linked to a URl parameter
        $modelInstances = [];

        // First, any bound models
        // Eg if route is /users/{id}, and (User $user) model is typehinted on method
        // If User model has `id` as an integer, then {id} param should be an integer
        $typeHintedEloquentModels = UrlParamsNormalizer::getTypeHintedEloquentModels($endpointData->method);
        foreach ($typeHintedEloquentModels as $argumentName => $modelInstance) {
            $routeKey = $modelInstance->getRouteKeyName();

            // Find the param name. In our normalized URL, argument $user might be param {user}, or {user_id}, or {id},
            if (isset($parameters[$argumentName])) {
                $paramName = $argumentName;
            } else if (isset($parameters["{$argumentName}_$routeKey"])) {
                $paramName = "{$argumentName}_$routeKey";
            } else if (isset($parameters[$routeKey])) {
                $paramName = $routeKey;
            } else {
                continue;
            }

            $modelInstances[$paramName] = $modelInstance;
        }

        // Next, non-Eloquent-bound parameters. They might still be Eloquent models, but model binding wasn't used.
        foreach ($parameters as $name => $data) {
            if (isset($data['type'])) continue;

            // If the url is /things/{id}, try to find a Thing model
            $urlThing = $this->getNameOfUrlThing($endpointData->uri, $name);
            if ($urlThing && ($modelInstance = $this->findModelFromUrlThing($urlThing))) {
                $modelInstances[$name] = $modelInstance;
            }
        }

        // Now infer.
        foreach ($modelInstances as $paramName => $modelInstance) {
            // If the routeKey is the same as the primary key in the database, use the PK's type.
            $routeKey = $modelInstance->getRouteKeyName();
            $type = $modelInstance->getKeyName() === $routeKey
                ? static::normalizeTypeName($modelInstance->getKeyType()) : 'string';

            $parameters[$paramName]['type'] = $type;

            try {
// CHANGED THE FOLLOWING LINES:
                $modelInstance = $modelInstance::first();
                $parameters[$paramName]['example'] = $modelInstance->$routeKey ?? null;
                $parameters[$paramName]['custom']['modelInstance'] = $modelInstance ?? null;
            } catch (Throwable) {
                $parameters[$paramName]['example'] = null;
            }

        }
        return $parameters;
    }
}
```

Then, in the other `GetFromFormRequestExtended` I did:
```php
use Illuminate\Foundation\Http\FormRequest as LaravelFormRequest;
use Illuminate\Routing\Route;
use Knuckles\Camel\Extraction\ExtractedEndpointData;
use Knuckles\Scribe\Extracting\Strategies\BodyParameters\GetFromFormRequest;
use Knuckles\Scribe\Tools\Globals;

class GetFromFormRequestExtended extends GetFromFormRequest
{
    public function __invoke(ExtractedEndpointData $endpointData, array $routeRules = []): ?array
    {
        return $this->getParametersFromFormRequestWithModelBinding($endpointData, $endpointData->route);
    }

    public function getParametersFromFormRequestWithModelBinding(ExtractedEndpointData $endpointData, Route $route): array
    {
        if (!$formRequestReflectionClass = $this->getFormRequestReflectionClass($endpointData->method)) {
            return [];
        }

        if (!$this->isFormRequestMeantForThisStrategy($formRequestReflectionClass)) {
            return [];
        }

        $className = $formRequestReflectionClass->getName();

        if (Globals::$__instantiateFormRequestUsing) {
            $formRequest = call_user_func_array(Globals::$__instantiateFormRequestUsing, [$className, $route, $endpointData->method]);
        } else {
            $formRequest = new $className;
        }

        // Set the route properly so it works for users who have code that checks for the route.
        /** @var LaravelFormRequest|DingoFormRequest $formRequest */
        $formRequest->setRouteResolver(function () use ($formRequest, $route, $endpointData) {
            // Also need to bind the request to the route in case their code tries to inspect current request
            $route = $route->bind($formRequest);

// ADDED THE FOREACH BELOW:
            foreach ($endpointData->urlParameters as $urlParameter) {
                $route->setParameter($urlParameter->name, data_get($urlParameter, 'custom.modelInstance', $urlParameter->example));
            }

            return $route;
        });

        $formRequest->server->set('REQUEST_METHOD', $route->methods()[0]);

        $parametersFromFormRequest = $this->getParametersFromValidationRules(
            $this->getRouteValidationRules($formRequest),
            $this->getCustomParameterData($formRequest)
        );

        return $this->normaliseArrayAndObjectParameters($parametersFromFormRequest);
    }
}
```

After that, it worked as intended.

**Comment by coreymcmahon** [comment]
Solution from @igorsgm above works, with one caveat (at least in our case). We also needed to disable endpoint URL normalisation using:

```php
Scribe::normalizeEndpointUrlUsing(fn ($url) => $url);
```

This was because the route parameter was being normalised from the model name (eg: `user`) to `id`, which meant `$typeHintedEloquentModels[$paramName]` was empty.

Another solution here would be to simply inject _all_ `$typeHintedEloquentModels` values as route parameters:

```php
foreach ($typeHintedEloquentModels as $name => $model) {
    $route->setParameter($name, $model::first());
}
```

... but I'm not sure if there would be other implications with that approach.

**Comment by quevlu** [comment]
I don't know if I had the samen problem as you. However, I created a way to bind dynamic params in Requests

```
if (class_exists(Scribe::class)) {
            Scribe::instantiateFormRequestUsing(function (
                string $formRequestClassName,
                Route $route,
                \ReflectionFunctionAbstract $method
            ) {
                $reflectionClass = new \ReflectionClass($formRequestClassName);
                $params = $reflectionClass->getConstructor()->getParameters();

                $bindingParams = [];

                foreach ($params as $param) {
                    $bindingParams[] = app($param->getType()->getName());
                }

                return new $formRequestClassName(...$bindingParams);
            });
        }
```
